File: to_parquet.R
library(arrow)
library(PNADcIBGE)
library(here)

std_path <- getwd()

layout_path <- paste(std_path, "/PNAD_data/input_PNADC_trimestral.txt", sep="")
lpath <- here(std_path, "PNAD_data", "input_PNADC_trimestral.txt")

make_path <- function(year, trimester) {
  raw_path <- here("PNAD_data", year, paste("PNADC_0", trimester, year, ".txt", sep=""))
  parquet_path <- here(std_path, "PNAD_data", year, paste("PNADC_0", trimester, year, ".parquet", sep=""))
  paths <- c(raw_path, parquet_path)
  paths
}

make_parquet <- function(year, trimester) {
  columns_to_keep = c("Ano", "Trimestre", "UF", "UPA",
                      "V1008",  # Número de seleção do domicílio
                      "V1014",  # Painel
                      "V1027",  # Peso do domicílio e das pessoas (sem calibração)
                      "V1028",  # Peso do domicílio e das pessoas (com calibração)
                      "V2003",  # Número de ordem
                      "V2005",  # Condição no domicílio
                      "V2007",  # Sexo
                      "V2008",  # Dia de nascimento
                      "V20081", # Mês de nascimento
                      "V20082", # Ano de nascimento 
                      "V2009",  # Idade do morador na data de referência
                      "V2010",  # Cor ou raça
                      "V3009A", # Qual foi o curso mais elevado que ... frequentou anteriormente?
                      "V4010",  # Código da ocupação (cargo ou função) (COD) -> Ver  "Composição dos Grupamentos Ocupacionais" e "Classificação de Ocupações para as Pesquisas Domiciliares – COD" em ANEXO de Notas Metodológicas
                      "V4012",  # Nesse trabalho, ... era: 
                      "V4013",  # Código da principal atividade desse negócio/empresa (CNAE)
                      "V4029",  # Nesse trabalho, ... tinha carteira de trabalho assinada ?
                      "V4040",  # Até o dia ... (último dia da semana de referência) fazia quanto tempo que ... estava nesse trabalho ?
                      "V4041",  # Código da ocupação (cargo ou função) (COD) -> Ver "Classificação nacional de ocupações para pesquisas domiciliares (COD) 2010"
                      "V4043",  # Nesse trabalho secundário, ... era 
                      "VD4001", # Condição em relação à força de trabalho na semana de referência para pessoas de 14 anos ou mais de idade
                      "VD4002", # Condição de ocupação na semana de referência para pessoas de 14 anos ou mais de idade
                      "VD4009", # Posição na ocupação e categoria do emprego do trabalho principal da semana de referência para pessoas de 14 anos ou mais de idade
                      "VD4016", # Rendimento mensal habitual do trabalho principal para pessoas de 14 anos ou mais de idade (apenas para pessoas que receberam em dinheiro, produtos ou mercadorias no trabalho principal)
                      "VD4017", # Rendimento mensal efetivo do trabalho principal para pessoas de 14 anos ou mais de idade (apenas para pessoas que receberam em dinheiro, produtos ou mercadorias no trabalho principal)
                      "VD4019", # Rendimento mensal habitual de todos os trabalhos para pessoas de 14 anos ou mais de idade (apenas para pessoas que receberam em dinheiro, produtos ou mercadorias em qualquer trabalho) 
                      "VD4020", # Rendimento mensal efetivo de todos os trabalhos para pessoas de 14 anos ou mais de idade (apenas para pessoas que receberam em dinheiro, produtos ou mercadorias em qualquer trabalho)
                      "VD4035", # Horas efetivamente trabalhadas na semana de referência em todos os trabalhos para pessoas de 14 anos ou mais de idade
                      ) 

  paths = make_path(year, trimester)

  df <- read_pnadc(paths[1], layout_path, columns_to_keep)
  print("Arquivo txt lido.")
  print("Escrevendo parquet")
  write_parquet(df, paths[2], compression = "snappy")
}

if ((sys.nframe() == 0) | (interactive() & sys.nframe() %/% 4 == 1)) {
  year <- as.integer(readline("Ano dos microdados : "))
  trimester <- as.integer(readline("Trimestre desejado: "))
  make_parquet(year, trimester)
}

File: to_parquet.py
from pathlib import Path
import re
import pandas as pd
import pyarrow.parquet as pq
import pyarrow as pa

layout_path = Path("PNAD_data/input_PNADC_trimestral.txt")

def make_data_paths(year, trimester):
    """
    Dado o ano e trimestre desejados, retorna os paths para o arquivo referente e 
    para o local do Parquet a ser gerado.
    
    Parêmatros:
        year (int): ano para o arquivo em questão, no formato XXXX (ex.: 2021)
        trimester (int): trimestre para o arquivo em questão, no formato X (ex.: 1)
    """
    
    raw_path = f"PNAD_data/{year}/PNADC_0{trimester}{year}.txt" # path do arquivo direto do PNAD
    parquet_path = f"PNAD_data/{year}/PNADC_0{trimester}{year}.parquet" # path para o arquivo resultante
    
    return Path(raw_path), Path(parquet_path)

pattern = r'@(\d+)\s+(\w+)\s+([^\s]+)\.'  # extração de início, varname e formatação
starts, widths, names = [], [], []

with open(layout_path, 'r', encoding='latin1') as file:
    for line in file:
        match = re.search(pattern, line)
        if match:
            start = int(match.group(1))
            var = match.group(2)
            fmt = match.group(3)

            # comprimento pode ser derivado a partir do rpóximo início
            starts.append(start)
            names.append(var)

widths = [starts[i+1] - starts[i] for i in range(len(starts)-1)]
widths.append(4000 - starts[-1] + 1)

if __name__ == "__main__":
    y = input("Ano dos microdados:")
    t = input("Trimestre desejado:")

    txt_file, parquet_file = make_data_paths(y, t)

    columns_to_keep = ["Ano", "Trimestre", "UF", "UPA", "V1008", "V1014", "V2003", "V2005", 
                    "V2007", "V2009", "VD4016", "VD4002", "VD4019", "VD4035", "V3009A", "V2010"]

    # separando o arquivo e lendo por chunks
    chunksize = 100_000
    dfs = []
    total_rows = 0 

    for i, chunk in enumerate(pd.read_fwf(txt_file, 
                                        widths=widths, 
                                        names=names, 
                                        dtype=str, 
                                        encoding='latin1', 
                                        chunksize=chunksize)):
        total_rows += len(chunk)
        print(f"Lidas {total_rows:,} linhas ate agora...")

        chunk = chunk[[c for c in columns_to_keep if c in chunk.columns]]
        dfs.append(chunk)

    # concatenar os chunks
    df = pd.concat(dfs, ignore_index=True)
            
    print(f"Loaded {len(df):,} rows with {len(df.columns)} columns.")
            
    table = pa.Table.from_pandas(df)
            
    pq.write_table(table, parquet_file, compression='snappy')

    print(f"Saved to {parquet_file}")